irods/at_client_exit.py

```diff
@@ -0,0 +1,43 @@
+import atexit
+import enum
+import logging
+
+logger = logging.getLogger(__name__)
+
+class priority(enum.Enum):
+    # Integer value determines relative order of execution among the stages.
+    BEFORE_PRC_CLEANUP = 10
+    DURING_PRC_CLEANUP = 20
+    AFTER_PRC_CLEANUP = 30
```

Comment:
I think this enum definition needs some attention.

`priority` is pretty generic. Is there a qualifier that can be used to make its purpose more narrow? If the name can be more specific, then that would likely remove the need to include `_PRC_` in the enum value names.

If `_PRC_` must be part of the enum value names, then I think they should be changed so that each of them share the same prefix (e.g. `PRC_CLEANUP_*`). Especially if imported into the global namespace.

Reply:
We could say `ExecutionPriority`, `ExecutionOrder`, or `ExecutionStage` ... ?

Reply:
Also, I'm not sold on PRC being part of the name. I'd be open to suggestions as I've struggle with this for some time.

Reply:
Thoughts?

```py
class CleanupPolicy(enum.Enum):
    BEFORE = 10,
    DURING = 20,
    AFTER  = 30,
```

If these define the "when", then perhaps `Policy` can be replaced by something that expresses that. 

Can you provide an example which shows usage of this? It doesn't need to be real/complete. It just needs to show how it's handled by the user of the library.

Reply:
The class name could be, but  wouldn't necessarily have to be, referred to within an application that imports the library, -- since the constants (`BEFORE`,...) have been elevated to the irods.at_client_exit  module namespace at top level..,   So, keeping the definition of CleanupPolicy as you've given it, and assuming we renamed nothing else and added the following line as the last line of the module:
```
__all__ = ['register', 'CleanupPolicy'] + [
           name for name,obj in vars().items() if type(obj) is CleanupPolicy]
```              
Then an application could do this:
```
# Import all stage constants and the registering function itself
from irods.at_client_exit import *

# These functions are to be run at exit:
def run_me_before_prc_cleanup(): 
  pass
def run_me_after_prc_cleanup(): 
  pass
def run_me_along_with_prc_cleanup( ):
  pass

register( BEFORE, run_me_before_prc_cleanup)

# Register our own exit function to be run within the stage (or phase, if you like) 
# as the PRC's cleanup of its own affairs:

register( DURING,run_me_along_with_prc_cleanup)

# Inconsistent usage here, but demonstrating we can also scope the stage constants
# to their original parent namespace:
register( CleanupPolicy.AFTER,run_me_after_prc_cleanup)

# ... further application code
```       
Admittedly, we'd probably want to rename the `register` function to be a bit more descriptive.
In the case of PRC-external things, whatever we register under "DURING"  probably will just before PRC cleanup, per LIFO ordering, but the main point of implementing this whole thing is the BEFORE and AFTER aspect of it. I doubt DURING would be used that much tbh.

Reply:
Should we consider shipping just before and after at first?

Reply:
> Should we consider shipping just before and after at first?

I think so. `DURING` may be difficult to reason about. Seems the user would need to know unnecessary details to use it correctly.

> The enum class name can be, but wouldn't necessarily need to be, referred to within the application that imports and uses irods.at_client_exit , since the constants (BEFORE,etc.) have been elevated to the module's top level namespace. That's the reason behind this line.

@d-w-moore Why do the enum values need to be elevated to the top level namespace? Is that a requirement? The enum class name provides context to help readers understand the interface. The user can always import the enum values if they prefer that, correct?

> Admittedly, we'd probably want to rename the register function to be a bit more descriptive.

Agreed. In fact, the enum class could be dropped in favor of multiple functions. For example _(these are just examples)_:
- `register_cleanup_callback_pre`
- `register_cleanup_callback_post`

> In the case of PRC-external things, whatever we register under "DURING" probably will get called just before PRC cleanup, per LIFO ordering, but the main point of implementing this whole thing is the BEFORE and AFTER aspect of it. I doubt DURING would be used that much tbh.

Agreed. That's a vote for removing `DURING` entirely. I think providing before and after control points is all that's needed.

Reply:
Would we still need to have `DURING` around in order for `BEFORE` and `AFTER` to have any meaning? Or will there be some mechanism for the internal PRC cleanup to define the `DURING` methods? Hopefully that question makes sense...

Reply:
DURING is already used internally by the PRC to schedule the cleaning of data objects, so I guess the answer is yes.  Plus, DURING as part of the enum is needed for proper sequencing of the three categories relative to each other.  I could try I guess to hide it from outside access, but I'm not sure doing so serves much of a purpose.  Are we thinking it's a security issue?

Reply:
The fact that DURING might not be used much is, I believe, not a sufficient reason for concealing it.  There are cases in which it  might be useful.
I thank you for bringing it up, it spurred me to think about relative order within "DURING" for applications that  might use it and therefore to add commit 95591b7145d19dd16329b9cf5c30bf1470a37915 , to make sure the registration of sessions' and data objects' automated cleanup is done at the time irods.session is imported.  This makes a lot of sense, because the applications in question will likely want their own DURING-phase callbacks to be done before the whole PRC is built down (per the LIFO thing), and now they can do so, while still being sure that their  BEFORE-phase  exit callbacks will be done even sooner.
I also think it quite harmless, and  convenient in this case, to raise the symbols to module level.  There are other precedents for this, like when we use inline namespaces in C++.

Reply:
After some discussion, the plan is to keep the existing implementation, but treat it as a low-level interface and wrap it in a simpler one i.e. present a high-level interface consisting of two functions _(function names not final)_:
  - `register_cleanup_callback_pre`
  - `register_cleanup_callback_post`

Reply:
> Other names
> 
> * `execute_before_lib_cleanup`
> * `execute_after_lib_cleanup`
> * `cleanup_action_pre`
> * `cleanup_action_post`

Yes, we should consider some of these.  In fact, let me know if we should have a great naming.... It seems I'm a bit blocked on what's appropriate to call things right now.

Reply:
Sure. Let's use this space for the great naming. Post your suggestions here and we'll discuss them next week.

Reply:
> Sure. Let's use this space for the great naming. Post your suggestions here and we'll discuss them next week.

The only suggestions I have currently are that:
   -  the notion / name of "Priority" be replace with "Stage" .  Because we have exactly three stages  functionally speaking: BEFORE, DURING and AFTER, and calling it priorities is actually more misleading, especially  when mixed and used synonymously with "priority" as it is now.  
   - If we prefer "phase" to "stage", that would be fine too.
   - As for the enums that define those Priorities/Stages/Phases/_Whatever-we-want-to-name-them_  -- currently going by the names` {BEFORE|DURING|AFTER}_PRC_CLEANUP` -- are appropriately named, I think.  I vote we keep those as they are now.

Reply:
i like `stage` over `priority` - agreed.

stage names... `{BEFORE|DURING|AFTER}_PRC_CLEANUP` seem fine to me - i don't think they are confusing/misleading and i think they're sufficiently specific.

high level function names... `execute_{before|after}_lib_cleanup` seems the most direct.  or consider replacing `lib` with `prc`.

Reply:
I think I like that but maybe discuss today... If we aren't using the word "register" in the names of those free functions, there's a chance some might assume things get executed in the course of the function call.

Reply:
Yes, adding this as a discush item.

Reply:
After some discussion, we've decided the following:

The free functions will be called:
- `execute_before_prc_cleanup`
- `execute_after_prc_cleanup`
- `register_for_before_prc_cleanup`
- `register_for_after_prc_cleanup`
- `register_for_execution_before_prc_cleanup` _(winner)_
- `register_for_execution_after_prc_cleanup` _(winner)_
- `register_for_execution_before_cleanup`
- `register_for_execution_after_cleanup`
- `irods.at_client_exit.register_for_execution_before_cleanup`
- `irods.at_client_exit.register_for_execution_after_cleanup`

The enum will be:
```py
class LibraryCleanupStage(enum.Enum):
    BEFORE = 10,
    DURING = 20,
    AFTER  = 30,
```